### PR TITLE
Change validation & collation protocol names to include genesis hash & fork id

### PR DESCRIPTION
**Pull request by dmitry-markin**
A follow-up PR to https://github.com/paritytech/polkadot/pull/5870 implementing suggestions from https://github.com/paritytech/substrate/issues/7746.

### CODE REVIEW

File: node/network/protocol/src/peer_set.rs
Comment (dmitry-markin) [inline node/network/protocol/src/peer_set.rs:131]: May be it is worth keeping this `"/polkadot/validation/1"` & `"/polkadot/collation/1"` for backward compatibility? And also in `get_protocol_label()` below.

File: node/network/protocol/src/peer_set.rs
Comment (bkchr) [inline node/network/protocol/src/peer_set.rs:31]:
```suggestion
const LEGACY_VALIDATION_PROTOCOL_V1: &str = "/polkadot/validation/1";
const LEGACY_COLLATION_PROTOCOL_V1: &str = "/polkadot/collation/1";
```

File: node/network/protocol/src/peer_set.rs
Comment (bkchr) [reply node/network/protocol/src/peer_set.rs:131]: I think it is fine. You are probably not really interested in prometheus if someone is still on the old version.

File: node/network/protocol/src/peer_set.rs
Comment (bkchr) [inline node/network/protocol/src/peer_set.rs:138]: This is a code comment and not documentation. Aka move into the function above the match.

File: node/network/protocol/src/peer_set.rs
Comment (bkchr) [inline node/network/protocol/src/peer_set.rs:139]: I'm not convinced that we need this. The metrics should be okay to only work with the "normal" label. Especially as we already use the label for one metric. Not sure why we need two different versions.

WDYT @eskimor

File: node/network/bridge/src/network.rs
Comment (bkchr) [inline node/network/bridge/src/network.rs:71]: Generating this every time we send a message is really a bad idea :see_no_evil:

**Comment by bkchr** [comment]
Left some initial comments, didn't yet reviewed all of it.

File: node/network/bridge/src/network.rs
Comment (dmitry-markin) [reply node/network/bridge/src/network.rs:71]: Will fix

Comment (dmitry-markin) [reply node/network/bridge/src/network.rs:71]: Well, this time the `PeerSet` protocol versions are hardcoded throughout the codebase. This makes using a map `(protocol, version) -> protocol_name` a possible source of runtime errors. I'm thinking of a way to safeguard us from using any other versions of protocols than the ones registered in `PeerSetProtocolNames::new()`.

The best I can think of is creating the `ValidationVersion` & `CollationVersion` enums and making sure we have registered all the versions listed there. We can then use these enum variants in, e.g., `send_validation_message_v1()`. Nevertheless, these enum variants must eventually be converted into numbers, and I can't think of any other way than implementing a custom trait for version to number conversion. This must be a custom trait to make sure we only use our enums for version referencing. Nevertheless, this won't catch the errors of using `ValidationVersion` in place of `CollationVersion` and vice versa.

Comment (dmitry-markin) [reply node/network/bridge/src/network.rs:71]: To be honest, I don't like the solution proposed above. We'd have enums for version representation in a couple of places, but we'd still have numbers in protocol/PeerSet-agnostic parts. Do you see a better alternative @bkchr @eskimor ?

Comment (bkchr) [reply node/network/bridge/src/network.rs:71]: Hmm this entire versioning story in the crate looks fishy :see_no_evil: If a peer is connected on version 1 and you would send him data on version 2, this wouldn't work :see_no_evil: 

> Nevertheless, these enum variants must eventually be converted into numbers, and I can't think of any other way than implementing a custom trait for version to number conversion

Why? To automate the generation of the protocol names? You could implement `From<ValidationVersion> for u32` and then call `validation_version.into()`. This all sounds a little bit "hacky", but as I said above the current situation of versioning is not great in the crate either.

Comment (dmitry-markin) [reply node/network/bridge/src/network.rs:71]: > You could implement From<ValidationVersion> for u32 and then call validation_version.into().

This won't allow us to support passing of both `ValidationVersion` and `CollationVersion` otherwise as `u32`, so an API user might end up passing numeric constants in places where we want to limit them only to supported versions. With a custom trait the user would be at least limited to passing `ValidationVersion` __or__ `CollationVersion`. (But still could pass one instead of the other.)

Comment (dmitry-markin) [reply node/network/bridge/src/network.rs:71]: Added protocol name lookup instead of generating it every time in https://github.com/paritytech/polkadot/pull/5876/commits/0ec120b363b8dc6d24f699ce33dc0038ffedd464

File: node/network/protocol/src/peer_set.rs
Comment (bkchr) [inline node/network/protocol/src/peer_set.rs:281]: I would remove this. As there is no requirement that the validation & collation protocol have the same protocol version.

File: node/network/protocol/src/peer_set.rs
Comment (bkchr) [reply node/network/protocol/src/peer_set.rs:287]: `ProtocolVersion` could also be anything. Nothing enforces that this is only something that is defined by these enums.

File: node/network/protocol/src/peer_set.rs
Comment (bkchr) [inline node/network/protocol/src/peer_set.rs:198]: These are not used anywhere.

File: node/network/protocol/src/peer_set.rs
Comment (bkchr) [reply node/network/protocol/src/peer_set.rs:287]: We could make `ProtocolVersion` a unit type wrapper that only implements `From` for `CollationVersion` and `ValidationVersion`

**Review by bkchr — COMMENTED** [review]
The protocol version handling still looks very fragile and has quite some wrong assumptions.

File: node/network/protocol/src/peer_set.rs
Comment (dmitry-markin) [reply node/network/protocol/src/peer_set.rs:198]: The only place they are used is https://github.com/paritytech/polkadot/pull/5876/files#diff-64aa50a62fe0bddebd18c842650162a2b1b86b0abda390cc325658bb07d5e222L778-R840

File: node/network/protocol/src/peer_set.rs
Comment (eskimor) [reply node/network/protocol/src/peer_set.rs:136]: I don't really mind. Seeing how many connections are on each version could be interesting though.

**Review by eskimor — APPROVED** [review]
Approved modulo nits by Basti.

This is starting to become clumsy: Passing a `PeerSet`, a `ProtocolVersion` and a `PeerSetProtocolNames` through the whole stack, just for doing protocol_names.get_name(peerset, version) at the bottom :grimacing:  I think with this only at runtime known naming that plain static enumeration stopped making sense. Anyhow the whole networking stack needs some love anyway, so resolving that can be done at some later point in time, together with additional refactoring.

**Comment by eskimor** [comment]
Oh and before merge, I would love to have a burnin on Versi. @tdimitrov can you take care of that. Only burnin on half the network, if everything keeps running we should be good.

**Comment by tdimitrov** [comment]
> Oh and before merge, I would love to have a burnin on Versi. @tdimitrov can you take care of that. Only burnin on half the network, if everything keeps running we should be good.

Sure, I'll do it. 

@dmitry-markin, I see you are adding some more changes. Would you drop a comment when the PR is ready for a burnin?

File: node/network/protocol/src/peer_set.rs
Comment (dmitry-markin) [reply node/network/protocol/src/peer_set.rs:136]: I kept these version-differentiated labels for now to be able to see how many connections are on each version.

File: node/network/protocol/src/peer_set.rs
Comment (dmitry-markin) [reply node/network/protocol/src/peer_set.rs]: "Main" protocol name (as opposed to negotiated protocol name) is needed, e.g., for peer sets bookkeeping, but I've specified validation & collation versions explicitly using new enums.

File: node/network/protocol/src/peer_set.rs
Comment (dmitry-markin) [reply node/network/protocol/src/peer_set.rs]: Made `ProtocolVersion` a "strongly typed" struct. This still doesn't eliminate the risk of misusing it like `ProtocolVersion(1)`, but I've replaced all the hardcoded protocol versions with `ValidationVersion::V1.into()` and `CollationVersion::V1.into()`. Hope that the PR is finally OK. @bkchr could you check it again, please?

**Review by bkchr — APPROVED** [review]

File: node/network/bridge/src/rx/mod.rs
Comment (dmitry-markin) [reply node/network/bridge/src/rx/mod.rs:443]: Tried. It looks like such generalized `Option` comparisons don't work in Rust now: https://github.com/rust-lang/rust/issues/20063.

File: node/network/bridge/src/rx/mod.rs
Comment (dmitry-markin) [reply node/network/bridge/src/rx/mod.rs:476]: The same as above, `PartialEq` isn't enough for `Option` comparison.

**Comment by dmitry-markin** [comment]
> @dmitry-markin, I see you are adding some more changes. Would you drop a comment when the PR is ready for a burnin?

@tdimitrov The PR is ready for a burnin.

**Pull request merged**
